### PR TITLE
Changed sorting for TurtleWriter.GenerateOutput(TurtleWriterContext c…

### PR DESCRIPTION
**Pull request by MischaVreeburg**
…ontext)

Altered the List<triple>.Sort to use a layered dictionary (dictionary(subjectnode, dictionary( predicatenode, dictionary (objectnode, list<triple>))))
 and sorting the keys for each dictionary using the FastNodeComparer. Then use a list sort on each of the list<triples>, which is probable superfluous but if there are more than one it should sort, using FullTripleComparer(new FastNodeComparer()).

Because each of the sets are much smaller the sorting will be much faster.

### CODE REVIEW

File: Libraries/dotNetRDF/Writing/TurtleWriter.cs [hunk 1]
```diff
@@ -182,21 +167,60 @@ private void GenerateOutput(TurtleWriterContext context)
...
+                int capacity = ts.Count;
+                Dictionary<INode, Dictionary<INode, Dictionary<INode, List<Triple>>>> sortHelperDictionary = new Dictionary<INode, Dictionary<INode, Dictionary<INode, List<Triple>>>>(capacity);
+                // Fill dictionary
+                foreach (Triple triple in ts)
+                {
+                    if (!sortHelperDictionary.ContainsKey(triple.Subject)) { sortHelperDictionary.Add(triple.Subject, new Dictionary<INode, Dictionary<INode, List<Triple>>>()); }
+
+                    if (!sortHelperDictionary[triple.Subject].ContainsKey(triple.Predicate)) { sortHelperDictionary[triple.Subject].Add(triple.Predicate, new Dictionary<INode, List<Triple>>()); }
+
+                    if (!sortHelperDictionary[triple.Subject][triple.Predicate].ContainsKey(triple.Object)) { sortHelperDictionary[triple.Subject][triple.Predicate].Add(triple.Object, new List<Triple>()); }
```
Comment (kal) [inline Libraries/dotNetRDF/Writing/TurtleWriter.cs]: I think this level of the dictionary is not really needed. Objects don't have to be output in sort order, we just need the ordering to allow grouping by subject and predicate.

File: Libraries/dotNetRDF/Writing/TurtleWriter.cs [hunk 2]
```diff
@@ -182,21 +167,60 @@ private void GenerateOutput(TurtleWriterContext context)
...
+                        foreach (var objectKey in objectKeys)
+                        {
+                            var triplesSubset = sortHelperDictionary[subjectKey][predicatekey][objectKey];
+                            if (triplesSubset.Count > 1)
+                            {
+                                triplesSubset.Sort(new FullTripleComparer(new FastNodeComparer()));
+                            }
+
+                            ts.AddRange(triplesSubset);
+                        }
```
Comment (kal) [inline Libraries/dotNetRDF/Writing/TurtleWriter.cs]: If we don't have the object level of the sortHelperDictionary then this inner loop isn't needed so you could just add all the triples for a given subject+predicate pair.

Comment (MischaVreeburg) [reply Libraries/dotNetRDF/Writing/TurtleWriter.cs]: I added this level to get the same sorting as with List.Sort. If you say we don't need it, I'll remove it.

**Comment by MischaVreeburg** [comment]
Well, we all need our holidays. Thanks for the review.

**Pull request merged**